docs: add useful info on YSQL online CREATE INDEX (#7689)
yugabyte/yugabyte-db: docs/content/latest/api/ysql/the-sql-language/statements/ddl_create_index.md

`CONCURRENTLY`, `USING method`, `COLLATE`, and `TABLESPACE` options are not yet supported.

When an index is created on a populated table, YugabyteDB automatically backfills the existing data into the index.
In most cases, this uses an online schema migration.
The following table explains some of the differences between creating an index online and not online.

| Condition | Online | Not online |
| :-------- | :----- | :--------- |
| Safe to do other DMLs during `CREATE INDEX`? | yes | no |
| Keeps other transactions alive during `CREATE INDEX`? | mostly | no |

> | parallelizes index loading? | yes | no |

Reviewer: ```suggestion
| Parallelizes index loading? | yes | no |
```

`CONCURRENTLY`, `USING method`, `COLLATE`, and `TABLESPACE` options are not yet supported.

> When an index is created on a populated table, YugabyteDB automatically

Reviewer: Overall comment: please don't break lines in markdown.

There are two good options in markdown:
* one line per paragraph, as you would with a word processor
* one line per sentence — if we all did this, diffs would be easier to parse, but we don't...

backfills the existing data into the index.  In most cases, this uses an online
schema migration.

> | | online | not online |

Reviewer: ```suggestion
| Condition | Online | Not online |
```

Reviewer: Also, this table needs an introduction. What are you showing here?

"The following table explains..."

> | --- | --- | --- |

Reviewer: ```suggestion
| :-------- | :----- | :--------- |
```

> | safe to do other DMLs during `CREATE INDEX`? | yes\* | no |

Reviewer: ```suggestion
| Safe to do other DMLs during `CREATE INDEX`? | yes\* | no |
```

> | keeps other transactions alive during `CREATE INDEX`? | mostly | no |

Reviewer: ```suggestion
| Keeps other transactions alive during `CREATE INDEX`? | mostly | no |
```

| parallelizes index loading? | yes | no |

To disable online schema migration for YSQL `CREATE INDEX`, set the flag
`ysql_disable_index_backfill=true` on **all** nodes and **both** master and
tserver.  To disable online schema migration for one `CREATE INDEX`, use
`CREATE INDEX NONCONCURRENTLY`.

If online `CREATE INDEX` fails, it likely failed in the backfill step.  In that
case, the index exists but is not usable.  Drop the index and try again.  If it
still doesn't work,

- did it time out?  Then, try increasing timeout flags:
  - master `ysql_index_backfill_rpc_timeout_ms`
  - tserver `ysql_wait_until_index_permissions_timeout_ms`
- did you get "backfill failed to connect to DB"?  Then, you may be hitting an

>   issue with authentication.  If the version is <2.4 (stable) or <2.5.2

Reviewer: Maybe what we want to say here is actually this: 
```suggestion
  issue with authentication. If you're on a stable version prior to v2.4, or a 2.5 version prior to 2.5.2,
```

Reviewer: Cases like this really point out the trouble with out stable-vs-latest scheme of even/odd numbering :(

>   (latest), online `CREATE INDEX` does not work with authentication enabled.

Reviewer: ```suggestion
  online `CREATE INDEX` does not work with authentication enabled.
```

Author: no, this is incorrect because 2.4 is safe.

Author: nevermind. it's relevant now with other changes.

>   - If the version is 2.5.1 (latest), then you can use `CREATE INDEX

Reviewer: Current latest is 2.5.3.1...
```suggestion
  - If the version is at least 2.5.1, then you can use `CREATE INDEX
```

Author: "(latest)" and "(stable)" emphasize the release lines since there are two.  Odds are under the latest line; evens are under the stable line.

Reviewer: My point here is that you're calling out a non-current version in the `latest` release line. Are you meaning to call out 2.5.1 specifically, or just "if you're using 2.5.x, you can use CREATE INDEX"?

Line 73, as well, is out of date as of yesterday's release. This is the trouble with calling out version numbers. :(

Reviewer: Updated the suggestion :)

    NONCONCURRENTLY` as a workaround.

>   - If the version is >=2.3 (latest), then you can set

Reviewer: ```suggestion
  - If the version is at least 2.3, then you can set
```

    `ysql_disable_index_backfill=false` as a workaround.
  - In any case, you can disable authentication (e.g. using `ysql_enable_auth`,
    `ysql_hba_conf`, or `ysql_hba_conf_csv`) as a workaround.
- did you get "duplicate key value" error?  Then, you have a unique constraint
  violation.

> Please file a GitHub issue if you still have issues.

Reviewer: Let's direct to Slack first, and GitHub second. Likely a faster response on Slack.
```suggestion
Please ask for help in our [community Slack](https://www.yugabyte.com/slack) or [file a GitHub issue](https://github.com/yugabyte/yugabyte-db/issues/new?title=Index+backfill+failure) if you still have issues.
```

To prioritize keeping other transactions alive during the index backfill, bump

> up flags

Reviewer: ```suggestion
up the following flags:
```

- master `index_backfill_wait_for_old_txns_ms`
- tserver `ysql_index_state_flags_update_delay_ms`

> \* - There is a slight chance that correctness is violated.  To reduce that

Reviewer: This note should go right below the table
